File: api/download.js
// api/download.js — AnimePahe + kwik.si direct MP4 download
// Debug: /api/download?title=X&ep=1&lang=sub&debug=1  → shows each step's raw result

const ANIMEPAHE_API = (process.env.ANIMEPAHE_URL || 'https://animepahe-api-liard.vercel.app').replace(/\/$/, '');
const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36';

// ── Kwik decryption (ported from anime-dl/anime-downloader kwik.py) ───────────
const CHAR_MAP = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/';

function getString(content, s1, s2) {
  const slice2 = CHAR_MAP.slice(0, s2);
  let acc = 0;
  const chars = content.split('').reverse();
  for (let n = 0; n < chars.length; n++) {
    const c = chars[n];
    acc += (/\d/.test(c) ? parseInt(c, 10) : 0) * Math.pow(s1, n);
  }
  let k = '';
  while (acc > 0) { k = slice2[Math.floor(acc % s2)] + k; acc = Math.floor(acc / s2); }
  return k || '0';
}

function decryptKwik(fullStr, key, v1, v2) {
  v1 = parseInt(v1, 10); v2 = parseInt(v2, 10);
  let result = '', i = 0;
  while (i < fullStr.length) {
    let s = '';
    while (fullStr[i] !== key[v2]) { s += fullStr[i++]; }
    for (let j = 0; j < key.length; j++) s = s.split(key[j]).join(String(j));
    result += String.fromCharCode(parseInt(getString(s, v2, 10), 10) - v1);
    i++;
  }
  return result;
}

async function getKwikDownloadUrl(kwikUrl) {
  const pageRes = await fetch(kwikUrl, {
    headers: { 'Referer': 'https://animepahe.ru/', 'User-Agent': UA }
  });
  if (!pageRes.ok) throw new Error(`kwik page status ${pageRes.status}`);
  const html = await pageRes.text();
  const cookies = pageRes.headers.get('set-cookie') || '';

  const m = html.match(/\("(\w+)",\d+,"(\w+)",(\d+),(\d+),\d+\)/);
  if (!m) throw new Error('kwik: obfuscated params pattern not found in page HTML');

  const decrypted = decryptKwik(m[1], m[2], m[3], m[4]);
  const actionM = decrypted.match(/action="([^"]+)"/);
  const tokenM  = decrypted.match(/value="([^"]+)"/);
  if (!actionM || !tokenM) throw new Error('kwik: form action or token not found after decryption');

  const dlRes = await fetch(actionM[1], {
    method: 'POST', redirect: 'manual',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Referer': kwikUrl, 'Cookie': cookies, 'User-Agent': UA
    },
    body: `_token=${encodeURIComponent(tokenM[1])}`
  });

  const location = dlRes.headers.get('location');
  if (!location) throw new Error(`kwik: POST returned ${dlRes.status}, no Location header`);
  return location;
}

// ── Helper: normalise AnimePahe API response to array ─────────────────────────
function toArr(v) {
  if (!v) return [];
  if (Array.isArray(v)) return v;
  if (Array.isArray(v.data)) return v.data;
  if (v.data && typeof v.data === 'object') return Object.values(v.data);
  if (typeof v === 'object') return Object.values(v);
  return [];
}

// ── Main handler ──────────────────────────────────────────────────────────────
module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  if (req.method === 'OPTIONS') return res.status(200).end();

  const { title, ep, lang = 'sub', debug } = req.query;
  if (!title || !ep) return res.status(400).json({ error: 'title and ep required' });

  const log = [];
  const step = (name, data) => { log.push({ step: name, data }); console.log(`[download] ${name}:`, JSON.stringify(data).slice(0, 200)); };

  try {
    const epNum = parseInt(ep, 10);
    const wantDub = lang === 'dub';

    // ── Step 1: Search ────────────────────────────────────────────────────────
    const sr = await fetch(`${ANIMEPAHE_API}/search?q=${encodeURIComponent(title)}`, {
      headers: { 'User-Agent': UA, 'Referer': 'https://animepahe.ru/' }
    });
    const searchRaw = await sr.json();
    const results = toArr(searchRaw);
    step('search', { status: sr.status, count: results.length, first: results[0] });

    if (!results.length) {
      if (debug) return res.json({ log, error: 'search returned no results' });
      return res.status(404).json({ error: 'Anime not found on AnimePahe' });
    }

    const anime = results[0];
    const animeSession = anime.session || anime.id;

    // ── Step 2: Episodes ──────────────────────────────────────────────────────
    const page = Math.ceil(epNum / 30);
    const er = await fetch(`${ANIMEPAHE_API}/episodes?session=${animeSession}&page=${page}`, {
      headers: { 'User-Agent': UA, 'Referer': 'https://animepahe.ru/' }
    });
    const epRaw = await er.json();
    const episodes = toArr(epRaw);
    step('episodes', { status: er.status, page, count: episodes.length, firstEp: episodes[0], lastEp: episodes[episodes.length - 1] });

    const episode = episodes.find(e =>
      parseInt(e.number, 10) === epNum ||
      parseInt(e.episode, 10) === epNum ||
      parseInt(e.ep, 10) === epNum
    );

    if (!episode) {
      if (debug) return res.json({ log, error: `ep ${epNum} not matched in ${episodes.length} results` });
      return res.status(404).json({ error: `Episode ${ep} not found` });
    }

    const epSession = episode.session || episode.id;
    step('episode-match', { episode, epSession });

    // ── Step 3: Sources ───────────────────────────────────────────────────────
    const srcr = await fetch(
      `${ANIMEPAHE_API}/sources?anime_session=${animeSession}&episode_session=${epSession}`, {
      headers: { 'User-Agent': UA, 'Referer': 'https://animepahe.ru/' }
    });
    const srcRaw = await srcr.json();
    const sources = toArr(srcRaw);
    step('sources', { status: srcr.status, count: sources.length, samples: sources.slice(0, 2) });

    if (debug) {
      // In debug mode, stop here and return everything so far
      return res.json({ log, sources_raw: srcRaw });
    }

    const byLang = sources.filter(s => wantDub ? s.audio === 'eng' : s.audio === 'jpn');
    const toProcess = byLang.length ? byLang : sources;
    if (!toProcess.length) return res.status(404).json({ error: 'No sources found' });

    // ── Step 4: Kwik extraction ───────────────────────────────────────────────
    const downloadLinks = [];
    for (const src of toProcess.slice(0, 4)) {
      const kwikUrl = src.url || src.kwik || src.source || src.link;
      const quality = String(src.quality || src.resolution || '?').replace(/p$/, '') + 'p';
      if (!kwikUrl) { step('kwik-skip', { reason: 'no url field', src }); continue; }
      try {
        const mp4 = await getKwikDownloadUrl(kwikUrl);
        downloadLinks.push({ quality, url: mp4 });
        step('kwik-ok', { quality, mp4: mp4.slice(0, 60) + '...' });
      } catch (e) {
        step('kwik-fail', { quality, kwikUrl, error: e.message });
      }
    }

    if (!downloadLinks.length) return res.status(404).json({ error: 'kwik extraction failed' });
    return res.status(200).json({ sources: downloadLinks });

  } catch (err) {
    console.error('[api/download] fatal:', err.message);
    if (debug) return res.json({ log, fatal: err.message });
    return res.status(500).json({ error: err.message });
  }
};

File: api/download.py
"""
api/download.py
Place this file at api/download.py in your GitHub repo.
Vercel will run it as a Python serverless function.
Uses cloudscraper to bypass Cloudflare on AnimePahe.
"""
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs, quote
import json, re, time, os

try:
    import cloudscraper
    HAS_CS = True
except ImportError:
    import urllib.request
    HAS_CS = False

ANIMEPAHE = "https://animepahe.ru"
KWIK_BASE = "https://kwik.si"
ANILIST   = "https://graphql.anilist.co"

UA = ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
      "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36")

def scraper():
    if HAS_CS:
        s = cloudscraper.create_scraper(
            browser={"browser": "chrome", "platform": "windows", "mobile": False})
        s.headers.update({"User-Agent": UA, "Accept-Language": "en-US,en;q=0.9"})
        return s
    import requests
    s = requests.Session()
    s.headers.update({"User-Agent": UA})
    return s

def anilist_title(ani_id):
    import urllib.request
    body = json.dumps({
        "query": "query($id:Int){Media(id:$id,type:ANIME){title{english romaji}}}",
        "variables": {"id": int(ani_id)}
    }).encode()
    req = urllib.request.Request(ANILIST, data=body,
                                 headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=10) as r:
        d = json.loads(r.read())
    t = d.get("data", {}).get("Media", {}).get("title", {})
    return t.get("english") or t.get("romaji")

def pahe_get(sc, path):
    r = sc.get(ANIMEPAHE + path, headers={"Referer": ANIMEPAHE + "/"}, timeout=20)
    r.raise_for_status()
    return r.json()

def resolve_kwik(kwik_url, sc):
    hdrs = {"Referer": ANIMEPAHE + "/", "Origin": ANIMEPAHE, "User-Agent": UA}
    r = sc.get(kwik_url, headers=hdrs, timeout=20)
    html = r.text
    # Submit form if anti-bot present
    tok = re.search(r'name="_token"\s+value="([^"]+)"', html)
    act = re.search(r'<form[^>]+action="([^"]+)"', html)
    if tok and act:
        pr = sc.post(act.group(1), data={"_token": tok.group(1)},
                     headers={**hdrs, "Referer": kwik_url,
                              "Content-Type": "application/x-www-form-urlencoded"},
                     timeout=20)
        html = pr.text
    for pat in [
        r"source:\s*['\"]([^'\"]+\.m3u8[^'\"]*)['\"]",
        r'"file"\s*:\s*"([^"]+\.m3u8[^"]*)"',
        r"(https?://[^\s\"'<>]+\.m3u8[^\s\"'<>]*)",
    ]:
        m = re.search(pat, html)
        if m:
            return m.group(1).replace("\\/", "/"), kwik_url
    return None, None

def resolve(ani_id, ep_num, lang):
    sc = scraper()

    # 1. AniList title
    title = anilist_title(ani_id)
    if not title:
        raise ValueError("Could not get title from AniList")

    # 2. Search AnimePahe
    data = pahe_get(sc, f"/api?m=search&q={quote(title)}")
    results = data.get("data", [])
    if not results:
        raise ValueError(f"'{title}' not found on AnimePahe")
    anime = next((x for x in results
                  if title.lower() in x.get("title", "").lower()), results[0])

    # 3. Episodes — fetch correct page
    ep_page = max(1, (ep_num - 1) // 30 + 1)
    epdata = pahe_get(sc, f"/api?m=release&id={anime['session']}"
                         f"&sort=episode_asc&page={ep_page}")
    eps = epdata.get("data", [])
    ep = next((e for e in eps
               if round(float(e.get("episode", 0))) == ep_num), None)
    if not ep and 0 < ep_num <= len(eps):
        ep = eps[ep_num - 1]
    if not ep:
        raise ValueError(f"Episode {ep_num} not found")

    # 4. Sources
    raw = pahe_get(sc, f"/api?m=links&id={anime['session']}"
                       f"&session={ep['session']}&p=kwik")
    sources = []
    for row in raw.get("data", []):
        for q, info in row.items():
            if isinstance(info, dict):
                info["quality"] = q
                sources.append(info)
    if not sources:
        raise ValueError("No sources found")
    best = sorted(sources,
                  key=lambda s: int(re.sub(r"\D","",str(s.get("quality","720"))or"0")))[-1]
    kwik = best.get("kwik") or best.get("url")
    if not kwik:
        raise ValueError("No kwik URL in sources")

    # 5. Resolve kwik → m3u8
    m3u8, referer = resolve_kwik(kwik, sc)
    if not m3u8:
        raise ValueError("Could not extract m3u8 from kwik page")

    return {
        "success": True,
        "title":   title,
        "episode": ep_num,
        "quality": best.get("quality", "720p"),
        "url":     m3u8,
        "referer": referer,
    }


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        qs = parse_qs(urlparse(self.path).query)
        ani_id  = (qs.get("aniId",  [""])[0] or qs.get("aniid",  [""])[0]).strip()
        ep_num  = int(qs.get("ep",  ["1"])[0] or 1)
        lang    = qs.get("lang", ["sub"])[0]
        player  = f"https://vidnest.fun/animepahe/{ani_id}/{ep_num}/{lang}"

        self.send_response(200)
        self.send_header("Content-Type",                "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Cache-Control",               "no-store")
        self.end_headers()

        if not ani_id:
            self.wfile.write(json.dumps({"error": "aniId required"}).encode())
            return
        try:
            result = resolve(ani_id, ep_num, lang)
            self.wfile.write(json.dumps(result).encode())
        except Exception as e:
            fallback = {"success": False, "playerUrl": player, "error": str(e)}
            self.wfile.write(json.dumps(fallback).encode())

    def log_message(self, *_):
        pass   # silence Vercel logs
